File: libs/dotnet/SquareSums/Tree.cs
using System;
using System.Collections.Generic;
// ReSharper disable ForCanBeConvertedToForeach

namespace SquareSums
{
    internal class Tree
    {
        private Node?[] _nodes;

        private Node GetOrCreate(int n)
        {
            var node = _nodes[n - 1];
            if (node == null)
            {
                node = new Node(n);
                _nodes[n - 1] = node;
            }

            return node;
        }

        public Tree(int n)
        {
            _nodes = new Node?[n];
        }

        public IReadOnlyList<Node?> Roots()
        {
            return _nodes;
        }

        public void AddPair(int head, int tail)
        {
            Node headNode = GetOrCreate(head);
            Node tailNode = GetOrCreate(tail);
            headNode.Add(tailNode);
        }
        
        public void SortPairsUsing(NodesSortingFacade sorting)
        {
            for (var i = 0; i < _nodes.Length; i++)
            {
                var n = _nodes[i];
                if (n == null)
                {
                    throw new Exception("Unexpected null node");
                }

                sorting.SortNodes(n.Pairs);
            }

            sorting.SortNodes(_nodes.AsSpan()!);
        }

        public bool FinalizePairsAndVerifyAllNodesHavePairs()
        {
            for (var i = 0; i < _nodes.Length; i++)
            {
                var node = _nodes[i];
                if (node == null)
                {
                    return false;
                }
                
                node.FinalizePairs();
                
                if (node.PairsCount == 0)
                {
                    return false;
                }
            }

            return true;
        }
    }
}

File: libs/rust/square-sums/src/tree.rs
use std::{rc::Rc};

use crate::{node::Node, nodes_sorting_trait::NodesSorting};

pub struct Tree {
    pub roots: Vec<Rc<Node>>
}

impl Tree {
    pub fn new(roots: Vec<Rc<Node>>) -> Tree {
        Tree{roots}
    }
}

pub struct TreeBuilder {
    nodes: Vec<Option<Rc<Node>>>
}

impl TreeBuilder {

    pub fn new(n: u32) -> TreeBuilder {
        TreeBuilder{
            nodes: vec![None; (n+1)
             as usize]
        }
    }

    pub fn add_pair(&mut self, head: u32, tail: u32) {
        let head_node = self.get_or_create(head);
        let tail_node = self.get_or_create(tail);
        head_node.add(Rc::downgrade(&tail_node));
    }

    pub fn build<S: NodesSorting>(&mut self, sorting: &S) -> Result<Option<Tree>, String> {
        let size = self.nodes.len();
        let mut nodes = Vec::with_capacity(size);

        for index in 1..size {
            let builder_node = &self.nodes[index];

            match builder_node {
                Some(existing_node) => {
                    existing_node.finalize()?;
                    nodes.push(existing_node.clone())
                },
                None => return Ok(None),
            }

        }

        sorting.sort_nodes(&mut nodes);
        
        Ok(Some(Tree::new(nodes)))
    }

    fn get_or_create(&mut self, n: u32) -> Rc<Node> {
        let index = n as usize;
        match &self.nodes[index] {
            Some(node) => node.clone(),
            None => {
                let node = Rc::new(Node::new(n));
                self.nodes[index] = Some(node.clone());
                node
            },
        }
    }
}